internal/services/topic.go

```diff
@@ -80,24 +80,41 @@ func (ts *TopicService) Reconcile() (TopicReconcileResult, error) {
 	topicMetadata := metadata[0]
 
 	if topicMetadata.Err == sarama.ErrUnknownTopicOrPartition {
+
 		// canary topic doesn't exist, going to create it
 		log.Printf("The canary topic %s doesn't exist\n", topicMetadata.Name)
-		if result.Assignments, err = ts.createTopic(len(brokers)); err != nil {
-			log.Printf("Error creating topic %s: %v", topicMetadata.Name, err)
-			return result, err
+		// topic is created if "dynamic" reassignment is enabled or the expected brokers are provided by the describe cluster
+		if ts.canaryConfig.ExpectedClusterSize == config.ExpectedClusterSizeDefault ||
+			ts.canaryConfig.ExpectedClusterSize == len(brokers) {
+
+			if result.Assignments, err = ts.createTopic(len(brokers)); err != nil {
+				log.Printf("Error creating topic %s: %v", topicMetadata.Name, err)
+				return result, err
+			}
+			log.Printf("The canary topic %s was created\n", topicMetadata.Name)
+		} else {
+			// not creating the topic and returning error to avoid starting producer/consumer
+			return result, topicMetadata.Err
 		}
-		log.Printf("The canary topic %s was created\n", topicMetadata.Name)
+
 	} else {
 		// canary topic already exists, check replicas assignments
 		log.Printf("The canary topic %s already exists\n", topicMetadata.Name)
 
-		result.RefreshMetadata = len(brokers) != len(topicMetadata.Partitions)
-		if result.Assignments, err = ts.alterTopic(len(topicMetadata.Partitions), len(brokers)); err != nil {
-			log.Printf("Error altering topic %s: %v", topicMetadata.Name, err)
-			return result, err
+		// topic partitions reassignment happens if "dynamic" reassignment is enabled
+		// or partitions are not equals to the expected brokers
+		if ts.canaryConfig.ExpectedClusterSize == config.ExpectedClusterSizeDefault ||
+			len(topicMetadata.Partitions) != ts.canaryConfig.ExpectedClusterSize {
```

Comment:
Just to be sure I understand, if we scale down then because Kafka doesn't support deleting partitions `len(partitions) > expectedSize`, so we reassign, but that's OK because we only run this at startup now, right?

Reply:
Reconcile is called at startup but even periodically.
Scaling down is obviously different from scale up. This condition works for scale up so reassignment is done only at startup even if the method is called more times but doesn't work for scale down where len(partitions) will be always higher than expected cluster size.
Good catch, have to fix this scenario.

Reply:
What I am trying to do is actually trying to reuse the same Reconcile() method to avoid code duplication across code at startup time and periodic code for reconcile.

Reply:
And actually the `len(topicMetadata.Partitions) != ts.canaryConfig.ExpectedClusterSize` doesn't work in case of scale down and scale up again when a reassignment should happen again but in this case won't.
I am simplifying the condition, I think that doing a reassignment if needed at the beginning is the simpler solution.